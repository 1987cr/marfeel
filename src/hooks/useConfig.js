// src/config-context.js
import React from "react";
import api from "api";
import logo from "assets/logo.svg";
import { black, white } from "styles/colors";

export const defaultConfig = {
  header: {
    background: white
  },
  logo: {
    url: logo,
    width: 147,
    height: 36
  },
  burger: {
    color: black
  },
  section: {
    color: black,
    size: 14
  },
  tabs: {
    home: "Home",
    politics: "Politics",
    sports: "Sports",
    national: "National",
    technology: "Tech"
  }
};

export const mergeConfig = (defaultConfig, config) =>
  Object.keys(defaultConfig).reduce(
    (prev, key) => ({
      ...prev,
      [key]: {
        ...defaultConfig[key],
        ...(config[key] || {})
      }
    }),
    {}
  );

const ConfigContext = React.createContext();

export const ConfigProvider = ({ children }) => {
  const [config, setConfig] = React.useState(null);

  React.useEffect(() => {
    const fetchConfig = async () => {
      try {
        const config = await api.fetchConfig();
        setConfig(mergeConfig(defaultConfig, config));
      } catch (error) {
        setConfig(defaultConfig);
      }
    };

    fetchConfig();
  }, []);

  return (
    <ConfigContext.Provider value={config}>
      {config ? children : null}
    </ConfigContext.Provider>
  );
};

const useConfig = () => {
  const context = React.useContext(ConfigContext);
  if (context === undefined) {
    throw new Error("useConfig must be used within a ConfigProvider");
  }
  return context;
};

export default useConfig;
