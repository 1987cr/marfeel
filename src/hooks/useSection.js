// src/section-context.js
import React from "react";
import api from "api";

const SectionStateContext = React.createContext();
const SectionDispatchContext = React.createContext();

export const STATUS = {
  LOADING: "loading",
  ERROR: "error",
  OK: "ok"
};

export const ACTION = {
  FETCH_START: "fetchStart",
  FETCH_SUCCESS: "fetchSuccess",
  FETCH_FAIL: "fetchFail",
  SET_SELECTED: "setSelected"
};

export const sectionReducer = (state, action) => {
  switch (action.type) {
    case ACTION.FETCH_START: {
      return {
        ...state,
        status: STATUS.LOADING
      };
    }
    case ACTION.FETCH_SUCCESS: {
      const articles = action.payload;

      return {
        ...state,
        status: STATUS.OK,
        articles: {
          ...state.articles,
          [state.selected]: articles
        }
      };
    }
    case ACTION.FETCH_FAIL: {
      return {
        ...state,
        status: STATUS.ERROR
      };
    }
    case ACTION.SET_SELECTED: {
      const selected = action.payload;

      return {
        ...state,
        selected
      };
    }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`);
    }
  }
};

export const initialState = {
  status: STATUS.OK,
  articles: {
    home: [],
    politics: [],
    sports: [],
    national: [],
    technology: []
  },
  selected: "home"
};

export const SectionProvider = ({ children }) => {
  const [state, dispatch] = React.useReducer(sectionReducer, initialState);

  const fetchSection = async section => {
    dispatch({ type: ACTION.FETCH_START });
    try {
      const articles = await api.fetchSection(section);
      dispatch({ type: ACTION.FETCH_SUCCESS, payload: articles });
    } catch {
      dispatch({ type: ACTION.FETCH_ERROR });
    }
  };

  const setSelected = section =>
    dispatch({ type: ACTION.SET_SELECTED, payload: section });

  const actions = { fetchSection, setSelected };

  return (
    <SectionStateContext.Provider value={state}>
      <SectionDispatchContext.Provider value={actions}>
        {children}
      </SectionDispatchContext.Provider>
    </SectionStateContext.Provider>
  );
};

const useSectionState = () => {
  const context = React.useContext(SectionStateContext);
  if (context === undefined) {
    throw new Error("useSectionState must be used within a SectionProvider");
  }
  return context;
};

const useSectionDispatch = () => {
  const context = React.useContext(SectionDispatchContext);
  if (context === undefined) {
    throw new Error("useSectionDispatch must be used within a SectionProvider");
  }
  return context;
};

const useSection = () => [useSectionState(), useSectionDispatch()];

export default useSection;
