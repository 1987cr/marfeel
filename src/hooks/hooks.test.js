import { useReducer } from "react";
import { renderHook, act } from "@testing-library/react-hooks";
import { sectionReducer, initialState, ACTION, STATUS } from "./useSection";
import { mergeConfig, defaultConfig } from "./useConfig";

describe("useSection: Section reducer", () => {
  it("throws if unknown action type is dispatched", () => {
    const { result } = renderHook(() =>
      useReducer(sectionReducer, initialState)
    );
    const [, dispatch] = result.current;

    const actionType = "INCREMENT";

    act(() => {
      dispatch({ type: actionType });
    });

    expect(result.error).toEqual(Error(`Unhandled action type: ${actionType}`));
  });

  it("status is set to 'loading' when fetch starts", () => {
    const { result } = renderHook(() =>
      useReducer(sectionReducer, initialState)
    );
    const [, dispatch] = result.current;

    act(() => {
      dispatch({ type: ACTION.FETCH_START });
    });

    const [state] = result.current;
    expect(state.status).toEqual(STATUS.LOADING);
  });

  it("status is set to 'error' when fetch fails", () => {
    const { result } = renderHook(() =>
      useReducer(sectionReducer, initialState)
    );
    const [, dispatch] = result.current;

    act(() => {
      dispatch({ type: ACTION.FETCH_FAIL });
    });

    const [state] = result.current;
    expect(state.status).toEqual(STATUS.ERROR);
  });

  it("articles are assigned to the correct list after fetch", () => {
    const { result } = renderHook(() =>
      useReducer(sectionReducer, initialState)
    );
    const [, dispatch] = result.current;

    const selected = "politics";
    const articles = [
      {
        id: 1,
        title: "Title 1",
        description: "Description 1"
      },
      {
        id: 1,
        title: "Title 2",
        description: "Description 2"
      }
    ];

    act(() => {
      dispatch({ type: ACTION.SET_SELECTED, payload: selected });
      dispatch({ type: ACTION.FETCH_SUCCESS, payload: articles });
    });

    const [state] = result.current;
    expect(state.selected).toEqual(selected);
    expect(state.articles[selected].length).toEqual(articles.length);
  });
});

describe("useConfig: Merge config", () => {
  it("result matchs snapshot", () => {
    const remoteConfig = {
      header: {
        background: "#000"
      },
      burger: {
        color: "#fff"
      },
      section: {
        size: 16,
        color: "#fff"
      },
      tabs: {
        technology: "Technology"
      }
    };
    const resultConfig = mergeConfig(defaultConfig, remoteConfig);
    expect(resultConfig).toMatchSnapshot();
  });
});
