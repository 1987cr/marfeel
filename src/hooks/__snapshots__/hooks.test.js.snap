// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useConfig: Merge config result matchs snapshot 1`] = `
Object {
  "burger": Object {
    "color": "#fff",
  },
  "header": Object {
    "background": "#000",
  },
  "logo": Object {
    "height": 36,
    "url": "logo.svg",
    "width": 147,
  },
  "section": Object {
    "color": "#fff",
    "size": 16,
  },
  "tabs": Object {
    "home": "Home",
    "national": "National",
    "politics": "Politics",
    "sports": "Sports",
    "technology": "Technology",
  },
}
`;
